gcc/rust/typecheck/rust-tyty.h

```diff
@@ -1071,35 +1071,74 @@ class ADTType : public BaseType, public SubstitutionRef
 class FnType : public BaseType, public SubstitutionRef
 {
 public:
+  // FIXME these could be constants
 #define FNTYPE_DEFAULT_FLAGS 0x00
 #define FNTYPE_IS_METHOD_FLAG 0x01
 #define FNTYPE_IS_EXTERN_FLAG 0x02
 #define FNTYPE_IS_VARADIC_FLAG 0X04
 
-  FnType (HirId ref, DefId id, std::string identifier, uint8_t flags,
+  enum ABI
+  {
+    UNKNOWN,
+    RUST,
+    INTRINSIC,
+    C,
+  };
+
+  static ABI get_abi_from_string (const std::string &abi, Location locus)
+  {
+    if (abi.compare ("rust") == 0)
+      return ABI::C;
+    else if (abi.compare ("rust-intrinsic") == 0)
+      return ABI::INTRINSIC;
+    else if (abi.compare ("C") == 0)
+      return ABI::C;
+
+    rust_error_at (locus, "unknown abi specified");
+    return ABI::UNKNOWN;
+  }
+
+  static std::string get_string_from_abi (ABI abi)
+  {
+    switch (abi)
+      {
+      case ABI::RUST:
+	return "rust";
+      case ABI::INTRINSIC:
+	return "rust-intrinsic";
+      case ABI::C:
+	return "C";
+
+      case ABI::UNKNOWN:
+	return "unknown";
+      }
+    return "unknown";
```

Comment:
```suggestion
      default:
	return "unknown";
      }
```

Reply:
Or are there any other ABI types for the future?

Reply:
There will be others in the future like rust-call, efiapi and architecture specific ones.

Reply:
yeah, we will be adding more here over time, rust-call is definitely the next one on the list. I am unsure how rustc handles the other architecture-specific ones, is this an LLVM feature?

Reply:
Rust has it's own set of abi's. Rustc maps some of them directly to LLVM abi's and has extra code for others. Most rust specific ones turn into the C abi at the LLVM side, but for example efiapi depends is either windows fastcall or the C abi depending on the architecture.